### ISSUE TITLE
[12.0]could not be found GeoIP on your system.

### ISSUE THREAD

**Issue by bizzappdev** [odoo/odoo#35524]
Impacted versions: 12.0

Everything is installed properly. all configurations are also done properly.
with default Odoo, I am getting this error.

If we add the `from odoo import _geoip_resolver` just before [this line](https://github.com/odoo/odoo/blob/12.0/addons/crm_reveal/models/crm_reveal_rule.py#L77) it is working properly.

IMO this can be a problem because the import of the _geoip_resolver done when Odoo started. and it has a default value. while the session wise value will be generated when the first call will be done from the webclient side based on IP.

![Selection_102](https://user-images.githubusercontent.com/4273984/62608955-7f745a80-b91e-11e9-99fc-a1312083196f.png)

**Comment by pedrobaeza** [comment]
I think a Python library is missing. Check warnings at the start of the service log

**Comment by bizzappdev** [comment]
@pedrobaeza  no that is not a problem. lib is already there and as I mentioned If I have the import statement just before the if condition then it is working.

**Comment by pedrobaeza** [comment]
Uhm, maybe some mangling with the way this is done in https://github.com/odoo/odoo/blob/cb70c8078128aed370d7566971e2c4979cb4708d/addons/http_routing/models/ir_http.py#L270

Have you tried putting http_routing as server wide module?

**Comment by bizzappdev** [comment]
@pedrobaeza  Yeah, still it was not working.

**Comment by pedrobaeza** [comment]
@Yenthe666 do you know who is in charge of this?

**Comment by Yenthe666** [comment]
@d-fence most likely! Could you have a look @d-fence?

**Comment by blaggacao** [comment]
> IMO this can be a problem because the import of the _geoip_resolver done when Odoo started. and it has a default value. while the session wise value will be generated when the first call will be done from the webclient side based on IP.

That is not true, within the global scope (server wide) it is global, yet lazily initialized. It must be a different error. Actually, you can check if initialization fails by asserting the NoneType just before it blows. It is initialized during "compile time" with `None` but with `False` if initialization failed.

This _can_ be a next step in finding the root cause. see [here](https://github.com/odoo/odoo/blob/cb70c8078128aed370d7566971e2c4979cb4708d/addons/http_routing/models/ir_http.py#L276). Hope it helps (a little further).

**Comment by blaggacao** [comment]
https://github.com/odoo/odoo/blob/12.0/addons/crm_reveal/models/crm_reveal_rule.py#L11
Possibly also:
```python
>>> import datetime as d
>>> from datetime import datetime
>>> datetime = None
>>> c.datetime
<class 'datetime.datetime'>
```

**Comment by bizzappdev** [comment]
@d

> > IMO this can be a problem because the import of the _geoip_resolver done when Odoo started. and it has a default value. while the session wise value will be generated when the first call will be done from the webclient side based on IP.
> 
> That is not true, within the global scope (server wide) it is global, yet lazily initialized. It must be a different error. Actually, you can check if initialization fails by asserting the NoneType just before it blows. It is ([eventually](https://github.com/odoo/odoo/blob/cb70c8078128aed370d7566971e2c4979cb4708d/addons/http_routing/geoipresolver.py#L16-L40)) initialized during "compile time" with `None` but with `False` if initialization failed.
> 
> This _can_ be a next step in finding the root cause. see [here](https://github.com/odoo/odoo/blob/cb70c8078128aed370d7566971e2c4979cb4708d/addons/http_routing/models/ir_http.py#L276). Hope it helps (a little further).
> 
> EDIT: you'd have to complement with checking the presence of [this log line](https://github.com/odoo/odoo/blob/cb70c8078128aed370d7566971e2c4979cb4708d/addons/http_routing/models/ir_http.py#L278).

I doubt your analysis, as it really looks like a problem of reference and value. 
when you import directly `from odoo import _geoip_resolver` and user `_geoip_resolver` directly it will behave and use the within local scope.

If you will use it `import odoo` and then use `odoo._geoip_resolver` then it will have a reference.  and in this case, odoo._geoip_resolver will have always the latest value.

the solution is  `import odoo`  and use from `odoo._geoip_resolver` at https://github.com/odoo/odoo/blob/12.0/addons/crm_reveal/models/crm_reveal_rule.py#L11

**Comment by blaggacao** [comment]
>I doubt your analysis, as it really looks like a problem of reference and value.
when you import directly from odoo import _geoip_resolver and user _geoip_resolver directly it will behave and use the within local scope.

Yeah, my first try was wrong. Thanks for pointing this out!

**Comment by blaggacao** [comment]
A challenging bug. Thanks! :stuck_out_tongue_winking_eye:

**Comment by bizzappdev** [comment]
Indeed. I wonder how it is working for others?

**Comment by madprog** [comment]
@bizzappdev It may work for others if the Odoo modules are loaded in a different order and one of the previously loaded modules makes `odoo._geoip_resolver` initialized before `crm_reveal` tries to access it.

**Comment by madprog** [comment]
Usually in Python, methods and attributes beginning with an underscore are considered as private and should not be accessed by "users".
Should we consider having a `get_geoip_resolver` function in the `odoo` module that would return `odoo._geoip_resolver` (and maybe initialize it if it was not initialized yet)?
That way, we would `from odoo import get_geoip_resolver` and call it when we need to use it.

**Comment by bizzappdev** [comment]
@madprog  that was also the solution is possible.
But what will be the use of having another method in odoo and calling it?
if we have always use odoo._geoip_resolver then also the problem is already solved. as it will remain as a reference and will have always the latest value. 
Will there be any other benefits for the same? let me know if I am missing something.

as in [_dispatch](https://github.com/odoo/odoo/blob/cb70c8078128aed370d7566971e2c4979cb4708d/addons/http_routing/models/ir_http.py#L340) method, it has been always resolved. it will be called on the very first call from client to server.

**Comment by madprog** [comment]
@bizzappdev If you try to access `odoo._geoip_resolver` before it was initialized by `ir_http`, then you will still find it set to `None`.
Instead, if you call the method that would initialize it if needed, then you would always get the after-initialization value. No need for the call in `_dispatch` any more, and all future developments would be able to use it, even if they do it before any `_dispatch` call for whatever reason.
However, if we are sure that this use case won't exist, #35625 looks OK to me.

**Comment by bizzappdev** [comment]
@madprog yeah True if it is not initialized by ir_http it will remain None.
But is there any chance that it will be used before _dispatch has been called? might be cron jobs?

**Comment by madprog** [comment]
Actually, it may be used by any custom module written by anyone, expecting that it was initialized. Usually, _dispatch would have been called before by any request, but yes, it could be in a cron job.

Imagine for example that a custom module wants to track where a received email came from, and retrieves the original sender's IP from the headers, and wants to use _geoip_resolver to determine the country of the sender.

**Comment by bizzappdev** [comment]
Very true, we should go your way for a new method `get_geoip_resolver` in the Odoo.
I can change the same PR for this.
what you think the best place where this new module should go? 
in  'ir.http' model itself? or it should be implemented as lib?

**Comment by blaggacao** [comment]
Very good argument to wrap this global (lazily loaded) sentinel into a propper accessor!

**Comment by blaggacao** [comment]
Can't we just du `@property` on the module level? (Outside of a class)

**Comment by blaggacao** [comment]
I think it should live in the `odoo` namespace itself, it indicates if a runtime library has been resolved, not necessarily _only_ a routing library as @madprog pointed out.

**Comment by sswapnesh** [comment]
> Indeed. I wonder how it is working for others?

There are plenty of things like this but Not everyone likes to Report/Improve.

**Comment by madprog** [comment]
It appears that a similar fix has been merged in cf42d8654917b4aac29bf5f8ba428ab6b516d741
This issue should probably be closed.

**Comment by sswapnesh** [comment]
This can be closed now.
Cc @pedrobaeza
